consensus/hotstuff/timeoutcollector/aggregation.go

```diff
@@ -0,0 +1,213 @@
+package timeoutcollector
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/verification"
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// signerInfo holds information about a signer, its public key and weight
+type signerInfo struct {
+	pk     crypto.PublicKey
+	weight uint64
+}
+
+// sigInfo holds signature and high QC view submitted by some signer
+type sigInfo struct {
+	sig        crypto.Signature
+	highQCView uint64
+}
+
+// TimeoutSignatureAggregator implements consensus/hotstuff.TimeoutSignatureAggregator.
+// It performs timeout specific BLS aggregation over multiple distinct messages.
+// We perform timeout signature aggregation for some concrete view, because of that our messages
+// are build in next way: hash(view, highestQC.View), where `highestQC` is a value contributed
+// by different replicas.
+// View and public keys needs to be agreed upon upfront.
+// Each signer is allowed to sign at most once.
+// Aggregation uses BLS scheme. Mitigation against rogue attacks is done using Proof Of Possession (PoP)
+// This module does not verify PoPs of input public keys, it assumes verification was done outside this module.
+// Implementation is thread-safe.
+type TimeoutSignatureAggregator struct {
+	lock          sync.RWMutex
+	hasher        hash.Hasher
+	idToInfo      map[flow.Identifier]signerInfo // auxiliary map to lookup signer weight and public key (only gets updated by constructor)
+	idToSignature map[flow.Identifier]sigInfo    // signatures indexed by the signer ID
+	totalWeight   uint64                         // total accumulated weight
+	view          uint64                         // view for which we are aggregating signatures
+	highestQC     *flow.QuorumCertificate        // highest QC submitted via VerifyAndAdd
```

Comment:
suggesting to track the newest QC on a higher code level. Technically, we could even include a QC that is newer than `TOHighQCViews`